File: java/net/meanscript/MSCode.java
package net.meanscript;
import net.meanscript.core.*;
import net.meanscript.java.*;
public class MSCode extends MC {
Common common;

MeanMachine mm;

boolean initialized;

public MSCode () throws MException
{
	common = new Common();
	mm = null;
	initialized = false;
	
	reset();
}
public MSCode (MSInputStream input, int streamType) throws MException
{
	MSJava.assertion(streamType >= MSJava.globalConfig.STREAM_TYPE_FIRST && streamType <= MSJava.globalConfig.STREAM_TYPE_LAST,   "unknown stream type");
	
	common = new Common();
	mm = null;
	initialized = false;
	
	if (streamType == MSJava.globalConfig.STREAM_BYTECODE)
	{
		initBytecode(input);
	}
	else if (streamType == MSJava.globalConfig.STREAM_SCRIPT)
	{
		mm = new MeanMachine(compile(input));
	}
	else
	{
		MSJava.assertion(false,   "unknown stream type");
	}
}

public MeanMachine getMM ()
{
	return mm;
}

//;

public void reset ()
{
	mm = null;
	initialized = false;
}

public boolean isInitialized ()
{
	return initialized;
}

public void checkInit () throws MException
{
	MSJava.assertion(initialized,   "MSCode is not initialized");
}

public void initBytecode (MSInputStream input) throws MException
{
	reset();
	
	ByteCode byteCode = new ByteCode(common, input);
	mm = new MeanMachine(byteCode);
	
	initialized = true;
}

public void initBytecode (ByteCode bc) throws MException
{
	reset();
	
	ByteCode byteCode = new ByteCode(bc);
	mm = new MeanMachine(byteCode);
	
	initialized = true;
}

public boolean hasData (String name) throws MException
{
	return mm.globals.hasData(name);
}

public boolean hasArray (String name) throws MException
{
	return mm.globals.hasArray(name);
}

public boolean getBool (String name) throws MException
{
	checkInit();
	return mm.globals.getBool(name);
}

public int getInt (String name) throws MException
{
	checkInit();
	return mm.globals.getInt(name);
}

public long getInt64 (String name) throws MException
{
	checkInit();
	return mm.globals.getInt64(name);
}

public float getFloat (String name) throws MException
{
	checkInit();
	return mm.globals.getFloat(name);
}
public double getFloat64 (String name) throws MException
{
	checkInit();
	return mm.globals.getFloat64(name);
}

public String getText (String name) throws MException
{
	checkInit();
	return mm.globals.getText(name);
}

public String getChars (String name) throws MException
{
	checkInit();
	return mm.globals.getChars(name);
}

public String getText (int textID) throws MException
{
	checkInit();
	return mm.globals.getText(textID);
}

public MSData getData (String name) throws MException
{
	checkInit();
	return mm.globals.getMember(name);
}

public MSDataArray getArray (String name) throws MException
{
	checkInit();
	return mm.globals.getArray(name);
}

public void writeReadOnlyData (MSOutputStream output) throws MException
{
	mm.writeReadOnlyData(output);
}

public void writeCode (MSOutputStream output) throws MException
{
	mm.writeCode(output);
}

public void printCode () throws MException
{
	if (initialized) mm.printCode();
	else MSJava.printOut.print("printCode: MSCode is not initialized").endLine();
}

public void printDetails () throws MException
{
	if (initialized) mm.printDetails();
	else MSJava.printOut.print("printDetails: MSCode is not initialized").endLine();
}

public void printData () throws MException
{
	if (initialized) mm.dataPrint();
	else MSJava.printOut.print("printDetails: MSCode is not initialized").endLine();
}

public ByteCode compile (MSInputStream input) throws MException
{
	reset();
	
	TokenTree tree = Parser.Parse (input);
	Semantics semantics = new Semantics();
	common.initialize(semantics);
	semantics.analyze(tree);
	ByteCode bc = Generator.generate (tree, semantics, common);
	tree = null;	
	
	initialized = true;
	return bc;
}

public void run () throws MException
{
	MSJava.assertion(mm != null,   "not initialized");
	mm.callFunction(0);
}

public void step () throws MException
{
	MSJava.assertion(false,   "TODO");
}

public void compileAndRun (String s) throws MException
{
	reset();
	MSInputArray ia = new MSInputArray(s);
	ByteCode bc = compile(ia);
	ia = null;
	mm = new MeanMachine(bc);
	run();
}

//public void compileAndRun (byte[] script) throws MException
//{
//	compile(script);
//	run();
//}


}

File: src/MeanscriptCpp/pub/MSCode.cpp
#include "MS.h"
namespace meanscript {
using namespace meanscriptcore;

MSCode::MSCode () 
{
	common = new Common();
	mm = 0;
	initialized = false;
	
	reset();
}
MSCode::MSCode (MSInputStream & input, int32_t streamType) 
{
	ASSERT(streamType >= globalConfig.STREAM_TYPE_FIRST && streamType <= globalConfig.STREAM_TYPE_LAST, "unknown stream type");
	
	common = new Common();
	mm = 0;
	initialized = false;
	
	if (streamType == globalConfig.STREAM_BYTECODE)
	{
		initBytecode(input);
	}
	else if (streamType == globalConfig.STREAM_SCRIPT)
	{
		mm = new MeanMachine(compile(input));
	}
	else
	{
		ASSERT(false, "unknown stream type");
	}
}

MeanMachine* MSCode::getMM ()
{
	return mm;
}

MSCode::~MSCode() { reset(); delete common; };

void MSCode::reset ()
{
	{ delete mm; mm = 0; };
	initialized = false;
}

bool MSCode::isInitialized ()
{
	return initialized;
}

void MSCode::checkInit () 
{
	ASSERT(initialized, "MSCode is not initialized");
}

void MSCode::initBytecode (MSInputStream & input) 
{
	reset();
	
	ByteCode* byteCode = new ByteCode(common, input);
	mm = new MeanMachine(byteCode);
	
	initialized = true;
}

void MSCode::initBytecode (ByteCode* bc) 
{
	reset();
	
	ByteCode* byteCode = new ByteCode((*bc));
	mm = new MeanMachine(byteCode);
	
	initialized = true;
}

bool MSCode::hasData (std::string name) 
{
	return (*(*mm).globals).hasData(name);
}

bool MSCode::hasArray (std::string name) 
{
	return (*(*mm).globals).hasArray(name);
}

bool MSCode::getBool (std::string name) 
{
	checkInit();
	return (*(*mm).globals).getBool(name);
}

int32_t MSCode::getInt (std::string name) 
{
	checkInit();
	return (*(*mm).globals).getInt(name);
}

int64_t MSCode::getInt64 (std::string name) 
{
	checkInit();
	return (*(*mm).globals).getInt64(name);
}

float MSCode::getFloat (std::string name) 
{
	checkInit();
	return (*(*mm).globals).getFloat(name);
}
double MSCode::getFloat64 (std::string name) 
{
	checkInit();
	return (*(*mm).globals).getFloat64(name);
}

std::string MSCode::getText (std::string name) 
{
	checkInit();
	return (*(*mm).globals).getText(name);
}

std::string MSCode::getChars (std::string name) 
{
	checkInit();
	return (*(*mm).globals).getChars(name);
}

std::string MSCode::getText (int32_t textID) 
{
	checkInit();
	return (*(*mm).globals).getText(textID);
}

MSData MSCode::getData (std::string name) 
{
	checkInit();
	return (*(*mm).globals).getMember(name);
}

MSDataArray MSCode::getArray (std::string name) 
{
	checkInit();
	return (*(*mm).globals).getArray(name);
}

void MSCode::writeReadOnlyData (MSOutputStream & output) 
{
	(*mm).writeReadOnlyData(output);
}

void MSCode::writeCode (MSOutputStream & output) 
{
	(*mm).writeCode(output);
}

void MSCode::printCode () 
{
	if (initialized) (*mm).printCode();
	else PRINT("printCode: MSCode is not initialized");
}

void MSCode::printDetails () 
{
	if (initialized) (*mm).printDetails();
	else PRINT("printDetails: MSCode is not initialized");
}

void MSCode::printData () 
{
	if (initialized) (*mm).dataPrint();
	else PRINT("printDetails: MSCode is not initialized");
}

void MSCode::dataOutputPrint (MSOutputPrint & output) 
{
	(*(*mm).globals).printData(output,0,"");
	output.close();
}

ByteCode* MSCode::compile (MSInputStream & input) 
{
	reset();
	
	TokenTree* tree = Parser::Parse (input);
	Semantics semantics = Semantics();
	(*common).initialize(semantics);
	semantics.analyze(tree);
	ByteCode* bc = Generator::generate (tree, semantics, (*common));
	{ delete tree; tree = 0; };	
	
	initialized = true;
	return bc;
}

void MSCode::run () 
{
	ASSERT(mm != 0, "not initialized");
	(*mm).callFunction(0);
}

void MSCode::step () 
{
	ASSERT(false, "TODO");
}

void MSCode::compileAndRun (std::string s) 
{
	reset();
	MSInputArray* ia = new MSInputArray(s);
	ByteCode* bc = compile((*ia));
	{ delete ia; ia = 0; };
	mm = new MeanMachine(bc);
	run();
}

//void MSCode::compileAndRun (uint8_t * script) 
//{
//	compile(script);
//	run();
//}


} // namespace meanscript(core)
